### DISCUSSION TITLE
mermaid js with Remix ESM

### DISCUSSION THREAD

**Discussion by angelinekwan** [remix-run/remix#5622] [Q&A] [answered]
Does anyone tried [mermaid](https://github.com/mermaid-js/mermaid) with Remix ? With the new [v10](https://github.com/mermaid-js/mermaid/blob/develop/CHANGELOG.md), it support ESM only. Failed to get it working properly might be due to [Remix to bundle the ESM module](https://remix.run/docs/en/1.13.0/pages/gotchas#importing-esm-packages) directly into the server build instead of requiring it at runtime causing the text node not matching between server and client side.

This code [example](https://codesandbox.io/s/react-with-mermaid-forked-4bly62?file=/src/index.js) works with basic react app setup but not in the context of Remix. 

Tested with remix indie-stack, code example below but it returns warning about hydration and switching to client side rendering. Even with the error, the diagram renders.

<img width="1370" alt="Screenshot 2023-03-01 at 19 10 19" src="https://user-images.githubusercontent.com/3796599/222214150-168d05dc-b6ba-4571-9636-6e147e7d7e63.png">

Error related to import mermaid which the node_modules has it installed.
`The path "mermaid" is imported in app/components/mermaid.tsx but "mermaid" was not found in your node_modules. Did you forget to install it?`

<img width="933" alt="Screenshot 2023-03-01 at 18 33 06" src="https://user-images.githubusercontent.com/3796599/222214044-9e4e8b76-9e9a-4606-9113-95b5ade878c2.png">

```
// components/mermaid.tsx
import React from "react";
import mermaid from "mermaid";

type MermaidProps = {
  children: string;
};

export function Mermaid({
  children,
}: MermaidProps) {

  React.useEffect(() => {
    const renderDiagram = async () => {
      mermaid.initialize({
        startOnLoad: false,
      });
      await mermaid.run({
        querySelector: ".mermaid",
        //suppressErrors: true,
      });
    };
    renderDiagram();
  }, [children]);
  
  if (!children) return null;

  return <div className="mermaid">{children}</div>;
};
```

```
// remix.config.js
/**
 * @type {import('@remix-run/dev').AppConfig}
 */
module.exports = {
  cacheDirectory: "./node_modules/.cache/remix",
  ignoredRouteFiles: ["**/.*", "**/*.css", "**/*.test.{js,jsx,ts,tsx}"],
  serverDependenciesToBundle: [
		/^mermaid.*/,
		/^khroma*/,
		/^d3*/,
		/^lodash-es*/,
		/^dagre-d3-es.*/,
		/^robust-predicates*/,
		/^internmap.*/,
	],
};
```

```
// app/routes/mermaid.tsx
import { Mermaid } from "~/components/mermaid";

export default function Index() {
  return (
    <>
    <p>Mermaid hardcoded 1</p>
			<Mermaid children={`
				graph TD;
					A-->B;
					A-->C;
					B-->D;
					C-->D;
			`} />
			<p>Mermaid hardcoded 2</p>
			<Mermaid children={`
				sequenceDiagram
				  Alice->>John: Hello John, how are you?
				  John-->>Alice: Great!
				  Alice-)John: See you later!		
			`} />
			<p>Mermaid hardcoded 3</p>
			<Mermaid children={`
				sequenceDiagram
				  Alice->>John: Hello John, how are you?
				  John-->>Alice: Great!
				  Alice-)John: See you later!		
			`} />
			<p>Mermaid hardcoded 4</p>
			<Mermaid children={`
				sequenceDiagram
				  Alice->>John: Hello John, how are you?
				  John-->>Alice: Great!
				  Alice-)John: See you later!		
			`} />
    </>
  );
}
```

**Comment by kiliman** [comment, marked as answer]
Here's a Remix sample with Mermaid

https://stackblitz.com/edit/remix-run-remix-76cte8?file=app%2Froutes%2Findex.tsx

**Reply by kiliman** [reply]
The more I look at this, the more it seems that these mermaid wrappers aren't really doing SSR and that's why you get the hydration errors.

**Reply by kiliman** [reply]
Wrapping the `<Mermaid>` component in `<ClientOnly>` from `remix-utils` will prevent the hydration errors. However, since mermaid generates static `<svg>` there's no reason it can't be SSR.

```ts
 return (
    <div>
      <ClientOnly>{() => <Mermaid chart={chart} />}</ClientOnly>
    </div>
  );
```

**Reply by angelinekwan** [reply]
Thanks @kiliman Your example works with the official mermaid as well
`import mermaid from "mermaid";`

**Comment by angelinekwan** [comment]
@kiliman Do you have an idea what this error might be related to.

In a more complex app setup, where the data is fetch externally via GraphQL, some mermaid diagram returns this error. [Source ](https://github.com/mermaid-js/mermaid/blob/6b251de227d999af7371e4928445193fe3a6de54/packages/mermaid/src/diagram-api/diagramAPI.ts#L68-L73)of the error in mermaid code.

<img width="1435" alt="Screenshot 2023-03-02 at 17 52 23" src="https://user-images.githubusercontent.com/3796599/222480776-723d5333-f17d-44de-85da-147d68ace719.png">

Checking the page element, those error mermaid diagram returns empty `<svg>`

<img width="1040" alt="Screenshot 2023-03-02 at 17 55 00" src="https://user-images.githubusercontent.com/3796599/222481158-eb655dc9-515a-4029-a61e-d1b2fdc076df.png">

**Reply by kiliman** [reply]
Do you have a minimal sample? Are you fetching GraphQL from your loaders or from the client?

**Reply by angelinekwan** [reply]
@kiliman Apparently the issue has been resolved. Might be due to an update on the packages [native mermaid 10.0.2](https://github.com/mermaid-js/mermaid/releases/tag/v10.0.2) 🤔 
<img width="1059" alt="Screenshot 2023-03-07 at 16 03 18" src="https://user-images.githubusercontent.com/3796599/223445346-79fa6bff-c604-4186-8253-0b5f3d0a816d.png">
